community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/helpers/TypeSafeMathSupport.scala

```diff
@@ -121,102 +128,112 @@ trait TypeSafeMathSupport {
+        case (l: Int, r: Float) => l - r
+        case (l: Int, r: Int) => l.toLong - r.toLong
+        case (l: Int, r: Long) => Math.subtractExact(l, r)
+        case (l: Int, r: Short) => l - r
+
+        case (l: Long, r: Byte) => Math.subtractExact(l, r)
+        case (l: Long, r: Double) => l - r
+        case (l: Long, r: Float) => l - r
+        case (l: Long, r: Int) => Math.subtractExact(l, r)
+        case (l: Long, r: Long) => Math.subtractExact(l, r)
+        case (l: Long, r: Short) => Math.subtractExact(l, r)
+
+        case (l: Short, r: Byte) => l - r
+        case (l: Short, r: Double) => l - r
+        case (l: Short, r: Float) => l - r
+        case (l: Short, r: Int) => l - r.toLong
+        case (l: Short, r: Long) => Math.subtractExact(l, r)
+        case (l: Short, r: Short) => l - r
 
+      }
+    } catch {
+      case e: java.lang.ArithmeticException  =>
+        throw new ArithmeticException(s"result of $left - $right cannot be represented as an integer")
     }
   }
 
   def multiply(left: Any, right: Any): Any = {
-    (left, right) match {
-      case (null, _) => null
-      case (_, null) => null
-
-      case (l: Byte, r: Byte)   => l * r
-      case (l: Byte, r: Double) => l * r
-      case (l: Byte, r: Float)  => l * r
-      case (l: Byte, r: Int)    => l * r
-      case (l: Byte, r: Long)   => l * r
-      case (l: Byte, r: Short)  => l * r
-
-      case (l: Double, r: Byte)   => l * r
-      case (l: Double, r: Double) => l * r
-      case (l: Double, r: Float)  => l * r
-      case (l: Double, r: Int)    => l * r
-      case (l: Double, r: Long)   => l * r
-      case (l: Double, r: Short)  => l * r
-
-      case (l: Float, r: Byte)   => l * r
-      case (l: Float, r: Double) => l * r
-      case (l: Float, r: Float)  => l * r
-      case (l: Float, r: Int)    => l * r
-      case (l: Float, r: Long)   => l * r
-      case (l: Float, r: Short)  => l * r
-
-      case (l: Int, r: Byte)   => l * r
-      case (l: Int, r: Double) => l * r
-      case (l: Int, r: Float)  => l * r
-      case (l: Int, r: Int)    => l * r
-      case (l: Int, r: Long)   => l * r
-      case (l: Int, r: Short)  => l * r
-
-      case (l: Long, r: Byte)   => l * r
-      case (l: Long, r: Double) => l * r
-      case (l: Long, r: Float)  => l * r
-      case (l: Long, r: Int)    => l * r
-      case (l: Long, r: Long)   => l * r
-      case (l: Long, r: Short)  => l * r
-
-      case (l: Short, r: Byte)   => l * r
-      case (l: Short, r: Double) => l * r
-      case (l: Short, r: Float)  => l * r
-      case (l: Short, r: Int)    => l * r
-      case (l: Short, r: Long)   => l * r
-      case (l: Short, r: Short)  => l * r
+    try {
+      (left, right) match {
+        case (null, _) => null
+        case (_, null) => null
+
+        case (l: Byte, r: Byte) => l * r
+        case (l: Byte, r: Double) => l * r
+        case (l: Byte, r: Float) => l * r
+        case (l: Byte, r: Int) => Math.multiplyExact(l, r.toLong)
```

Comment:
Wouldn't this always fit in a long? 2^8 * 2^32 = 2^40 < 2^64? And the same for some of the below operations, that doesn't include a `long`, or am I missing something?